### ISSUE TITLE
Odd Behaviour with QUINN and Apple Network Framework

### ISSUE THREAD

**Issue by paxsonsa** [quinn-rs/quinn#1742]
Thanks for the great library, docs, and for all the work you do. Thanks in advance for having a peek. Trying to track down an interesting problem; I have a small example here of a Quinn server running on macOS (13.6.3) with the latest version of Quinn and rust.  

The core issue I am running into is when the server is the one accepting the bidirectional stream after the tunnel is established, no data is received, however if I change the connection from accept_bi and open_bi the stream works perfectly as expected.

Am I missing something obvious here? Happy to provide the simplified swift server, figured most people might not want that.

Example:
```rust
/*
rustc 1.75.0 (82e1608df 2023-12-21)
Cargo.toml
quinn = "*"
bytes = "1.5"
rustls = { version = "0.21", features = ["quic"] }
rcgen = "0.12.0"
tokio = { version = "1.35.1", features = ["full"] }
thiserror = "1.0"
tracing = "0.1"
tracing-subscriber = "0.3"
*/
use bytes::BytesMut;
use quinn::{Endpoint, ServerConfig};
use std::str;
use std::{net::SocketAddr, sync::Arc};

pub const ALPN_QUIC_HTTP: &[&[u8]] = &[b"demo"];

#[tokio::main]
async fn main() {
    // Set up the tracing subscriber
    tracing::subscriber::set_global_default(
        tracing_subscriber::FmtSubscriber::builder()
            .with_env_filter(tracing_subscriber::EnvFilter::from_default_env())
            .finish(),
    )
    .unwrap();

    // TODO: Create Simple Server to Echo Messages from Client
    // - It would appear the Swift QUIC in Networking is not capable of customized CA
    // this is an issue. That might mean we need to use Rust Binding with Swift?
    // Another option is to message the Apple Help.
    let addr = "127.0.0.1:4567".parse().unwrap();
    run_server(addr).await;
    // tokio::spawn(run_server(addr));
    // run_client(addr).await.unwrap();
}

async fn run_server(addr: SocketAddr) {
    let cert = rcgen::generate_simple_self_signed(vec![
        "127.0.0.1".into(),
        "localhost".into(),
        "0.0.0.0".into(),
    ])
    .unwrap();
    let cert_der = cert.serialize_der().unwrap();
    let private_key = cert.serialize_private_key_der();
    let private_key = rustls::PrivateKey(private_key);
    let cert_chain = vec![rustls::Certificate(cert_der.clone())];

    let mut server_crypto = rustls::ServerConfig::builder()
        .with_safe_defaults()
        .with_no_client_auth()
        .with_single_cert(cert_chain, private_key)
        .unwrap();
    server_crypto.alpn_protocols = ALPN_QUIC_HTTP.iter().map(|&x| x.into()).collect();

    let server_config = ServerConfig::with_crypto(Arc::new(server_crypto));
    let endpoint = Endpoint::server(server_config, addr).unwrap();

    let incoming = endpoint.accept().await.unwrap();
    let conn = incoming.await.unwrap();
    println!(
        "[server] connection accepted: addr={}",
        conn.remote_address()
    );
    conn.set_max_concurrent_bi_streams(16_u8.into());

    // (A) This does not not work **********************************************************************************************************************************
    println!("waiting for bi stream");
    let (mut send, mut recv) = conn.accept_bi().await.unwrap();

    // (B) This works! *********************************************************************************************************************************************
    // let (mut send, mut recv) = conn.open_bi().await.unwrap();
    // send.write_all(b"hello").await.unwrap(); // Need this to open the stream on the client.

    println!("reading from recv");
    let mut header_bytes = bytes::BytesMut::zeroed(8);

    // ***************************************************************************************************************************************************************
    // (C) TIMEOUT HAPPENS HERE
    if let Err(err) = recv.read_exact(&mut header_bytes).await {
        println!("error reading header: {:?}", err);
        return;
    }
    println!("read bytes {:?}", header_bytes.len());
    let version = header_bytes[0];
    let message_type = header_bytes[1];
    let length = u32::from_be_bytes(header_bytes[2..6].try_into().unwrap());

    println!(
        "version: {:?}, message_type: {:?}, length: {:?}",
        version, message_type, length
    );

    // Read the body
    let mut body_bytes = BytesMut::zeroed(length as usize);
    recv.read(&mut body_bytes).await.unwrap();

    // Convert BytesMut to a &str, handling potential UTF-8 errors
    match str::from_utf8(&body_bytes) {
        Ok(s) => println!("String: {}", s),
        Err(e) => println!("Error decoding UTF-8: {}", e),
    }
}

// Implementation of `ServerCertVerifier` that verifies everything as trustworthy.
struct SkipServerVerification;

impl SkipServerVerification {
    fn new() -> Arc<Self> {
        Arc::new(Self)
    }
}

impl rustls::client::ServerCertVerifier for SkipServerVerification {
    fn verify_server_cert(
        &self,
        _end_entity: &rustls::Certificate,
        _intermediates: &[rustls::Certificate],
        _server_name: &rustls::ServerName,
        _scts: &mut dyn Iterator<Item = &[u8]>,
        _ocsp_response: &[u8],
        _now: std::time::SystemTime,
    ) -> Result<rustls::client::ServerCertVerified, rustls::Error> {
        Ok(rustls::client::ServerCertVerified::assertion())
    }
}
fn configure_client() -> quinn::ClientConfig {
    let crypto = rustls::ClientConfig::builder()
        .with_safe_defaults()
        .with_custom_certificate_verifier(SkipServerVerification::new())
        .with_no_client_auth();

    quinn::ClientConfig::new(Arc::new(crypto))
}
```

- (A) is where I am establishing stream. With (A) in place the receive at (C) times out. 
- If I comment out (A) and enable (B) there is no timeout at C, the recv returns.

**Interesting Behaviour?**
I at first thought, this must be a client issue with how I configured the Swift client (using Network Framework) but turning on trace logging for the server and battering the accepted stream with 20 byte packets I can see that the server is at least getting the bytes:
```
2024-01-16T08:07:12.538039Z TRACE first recv:frame{ty=CRYPTO}: quinn_proto::connection: consumed 245 CRYPTO bytes
2024-01-16T08:07:12.550967Z TRACE first recv: quinn_proto::connection: Handshake keys ready
2024-01-16T08:07:12.551143Z TRACE first recv: quinn_proto::connection: wrote 90 Initial CRYPTO bytes
2024-01-16T08:07:12.551336Z TRACE first recv: quinn_proto::connection: Data keys ready
2024-01-16T08:07:12.551414Z TRACE first recv: quinn_proto::connection: wrote 612 Handshake CRYPTO bytes
2024-01-16T08:07:12.551591Z TRACE quinn_proto::endpoint: connection incoming id=0 icid=b63dc15c23b8aa60
2024-01-16T08:07:12.551892Z TRACE drive{id=0}:send{space=Initial pn=0}: quinn_proto::connection: ACK ArrayRangeSet([0..1]), Delay = 0us
2024-01-16T08:07:12.551931Z TRACE drive{id=0}:send{space=Initial pn=0}: quinn_proto::connection: CRYPTO: off 0 len 90
2024-01-16T08:07:12.552539Z TRACE drive{id=0}:send{space=Handshake pn=0}: quinn_proto::connection: CRYPTO: off 0 len 612
2024-01-16T08:07:12.552597Z TRACE drive{id=0}:send{space=Handshake pn=0}: quinn_proto::connection::packet_builder: PADDING * 416
2024-01-16T08:07:12.552642Z TRACE drive{id=0}: quinn_proto::connection: sending 1200 bytes in 1 datagrams
2024-01-16T08:07:12.552816Z TRACE drive{id=0}:send{space=Data pn=0}: quinn_proto::connection: NEW_CONNECTION_ID sequence=1 id=6e28ba8144b2ba56
2024-01-16T08:07:12.552831Z TRACE drive{id=0}:send{space=Data pn=0}: quinn_proto::connection: NEW_CONNECTION_ID sequence=2 id=5c889e533bc89d9b
2024-01-16T08:07:12.552843Z TRACE drive{id=0}:send{space=Data pn=0}: quinn_proto::connection: NEW_CONNECTION_ID sequence=3 id=7af61158fa900620
2024-01-16T08:07:12.552853Z TRACE drive{id=0}:send{space=Data pn=0}: quinn_proto::connection: NEW_CONNECTION_ID sequence=4 id=755200bcbf4e520a
2024-01-16T08:07:12.552863Z TRACE drive{id=0}:send{space=Data pn=0}: quinn_proto::connection: NEW_CONNECTION_ID sequence=5 id=251eb4536d5d507c
2024-01-16T08:07:12.552874Z TRACE drive{id=0}:send{space=Data pn=0}: quinn_proto::connection: NEW_CONNECTION_ID sequence=6 id=eb302c48b048da45
2024-01-16T08:07:12.552884Z TRACE drive{id=0}:send{space=Data pn=0}: quinn_proto::connection: NEW_CONNECTION_ID sequence=7 id=f511776f08b5c241
2024-01-16T08:07:12.552905Z TRACE drive{id=0}: quinn_proto::connection: sending 214 bytes in 1 datagrams
2024-01-16T08:07:12.554476Z TRACE drive{id=0}: quinn_proto::connection: got Initial packet (1200 bytes) from 127.0.0.1:60003 using id a4ec05752f480a88
2024-01-16T08:07:12.554575Z DEBUG drive{id=0}:recv{space=Initial pn=1}:frame{ty=ACK}: quinn_proto::connection: ECN not acknowledged by peer
2024-01-16T08:07:12.554741Z TRACE drive{id=0}: quinn_proto::connection: got Handshake packet (39 bytes) from 127.0.0.1:60003 using id a4ec05752f480a88
2024-01-16T08:07:12.554763Z TRACE drive{id=0}:recv{space=Handshake pn=0}: quinn_proto::connection: discarding Initial keys
2024-01-16T08:07:12.554803Z TRACE drive{id=0}:recv{space=Handshake pn=0}: quinn_proto::connection: handshake ongoing
2024-01-16T08:07:12.554862Z TRACE drive{id=0}: quinn_proto::connection: got Handshake packet (73 bytes) from 127.0.0.1:60003 using id a4ec05752f480a88
2024-01-16T08:07:12.554937Z TRACE drive{id=0}:recv{space=Handshake pn=1}:frame{ty=CRYPTO}: quinn_proto::connection: consumed 36 CRYPTO bytes
2024-01-16T08:07:12.555075Z TRACE drive{id=0}:recv{space=Handshake pn=1}: quinn_proto::connection: wrote 324 Data CRYPTO bytes
2024-01-16T08:07:12.555090Z TRACE drive{id=0}:recv{space=Handshake pn=1}: quinn_proto::connection: discarding Handshake keys
2024-01-16T08:07:12.555106Z TRACE drive{id=0}:recv{space=Handshake pn=1}: quinn_proto::connection: established
2024-01-16T08:07:12.555140Z TRACE drive{id=0}:send{space=Data pn=1}: quinn_proto::connection: CRYPTO: off 0 len 324
2024-01-16T08:07:12.555166Z TRACE drive{id=0}: quinn_proto::connection: sending 347 bytes in 1 datagrams
2024-01-16T08:07:12.555186Z TRACE drive{id=0}:send{space=Data pn=2}: quinn_proto::connection::packet_builder: PADDING * 1307
2024-01-16T08:07:12.555206Z TRACE drive{id=0}: quinn_proto::connection: writing MTUD probe probe_size=1326
2024-01-16T08:07:12.555214Z TRACE drive{id=0}: quinn_proto::connection: sending 1326 bytes in 1 datagrams
[server] connection accepted: addr=127.0.0.1:60003
waiting for bi stream
2024-01-16T08:07:12.557409Z TRACE drive{id=0}: quinn_proto::connection: got Data packet (31 bytes) from 127.0.0.1:60003 using id a4ec05752f480a88
2024-01-16T08:07:12.557483Z TRACE drive{id=0}:recv{space=Data pn=0}: quinn_proto::connection: got frame Ack(Ack { largest: 2, delay: 0, ecn: None, ranges: "[0..=2]" })
2024-01-16T08:07:12.557528Z TRACE drive{id=0}:recv{space=Data pn=0}:frame{ty=ACK}: quinn_proto::connection::mtud: new MTU detected current_mtu=1326
2024-01-16T08:07:12.557576Z TRACE drive{id=0}:send{space=Data pn=3}: quinn_proto::connection::packet_builder: PADDING * 1370
2024-01-16T08:07:12.557600Z TRACE drive{id=0}: quinn_proto::connection: writing MTUD probe probe_size=1389
2024-01-16T08:07:12.557608Z TRACE drive{id=0}: quinn_proto::connection: sending 1389 bytes in 1 datagrams
2024-01-16T08:07:12.558793Z TRACE drive{id=0}: quinn_proto::connection: got Data packet (49 bytes) from 127.0.0.1:60003 using id a4ec05752f480a88
2024-01-16T08:07:12.558839Z TRACE drive{id=0}:recv{space=Data pn=1}: quinn_proto::connection: got stream frame id=client bidirectional stream 1 offset=0 len=20 fin=false
2024-01-16T08:07:12.558890Z TRACE drive{id=0}: quinn_proto::connection: got Data packet (50 bytes) from 127.0.0.1:60003 using id a4ec05752f480a88
2024-01-16T08:07:12.558916Z TRACE drive{id=0}:recv{space=Data pn=2}: quinn_proto::connection: got stream frame id=client bidirectional stream 1 offset=20 len=20 fin=false
2024-01-16T08:07:12.558942Z TRACE drive{id=0}: quinn_proto::connection: got Data packet (50 bytes) from 127.0.0.1:60003 using id a4ec05752f480a88
2024-01-16T08:07:12.558964Z TRACE drive{id=0}:recv{space=Data pn=3}: quinn_proto::connection: got stream frame id=client bidirectional stream 1 offset=40 len=20 fin=false
2024-01-16T08:07:12.558986Z TRACE drive{id=0}: quinn_proto::connection: got Data packet (50 bytes) from 127.0.0.1:60003 using id a4ec05752f480a88
2024-01-16T08:07:12.559007Z TRACE drive{id=0}:recv{space=Data pn=4}: quinn_proto::connection: got stream frame id=client bidirectional stream 1 offset=60 len=20 fin=false
2024-01-16T08:07:12.559028Z TRACE drive{id=0}: quinn_proto::connection: got Data packet (51 bytes) from 127.0.0.1:60003 using id a4ec05752f480a88
....
2024-01-16T08:07:15.763210Z TRACE drive{id=0}:recv{space=Data pn=18796}: quinn_proto::connection: got stream frame id=client bidirectional stream 1 offset=1244240 len=1434 fin=false
2024-01-16T08:07:15.763229Z TRACE drive{id=0}: quinn_proto::connection: got Data packet (1469 bytes) from 127.0.0.1:60003 using id a4ec05752f480a88
2024-01-16T08:07:15.763324Z TRACE drive{id=0}:recv{space=Data pn=18797}: quinn_proto::connection: got stream frame id=client bidirectional stream 1 offset=1245674 len=1434 fin=false
2024-01-16T08:07:15.763344Z TRACE drive{id=0}: quinn_proto::connection: got Data packet (1469 bytes) from 127.0.0.1:60003 using id a4ec05752f480a88
2024-01-16T08:07:15.763364Z TRACE drive{id=0}:recv{space=Data pn=18798}: quinn_proto::connection: got stream frame id=client bidirectional stream 1 offset=1247108 len=1434 fin=false
2024-01-16T08:07:15.763466Z TRACE drive{id=0}: quinn_proto::connection: got Data packet (1469 bytes) from 127.0.0.1:60003 using id a4ec05752f480a88
2024-01-16T08:07:15.763487Z TRACE drive{id=0}:recv{space=Data pn=18799}: quinn_proto::connection: got stream frame id=client bidirectional stream 1 offset=1248542 len=1434 fin=false
2024-01-16T08:07:15.763506Z TRACE drive{id=0}: quinn_proto::connection: got Data packet (58 bytes) from 127.0.0.1:60003 using id a4ec05752f480a88
2024-01-16T08:07:15.763525Z TRACE drive{id=0}:recv{space=Data pn=18800}: quinn_proto::connection: got stream frame id=client bidirectional stream 1 offset=1249976 len=24 fin=false
2024-01-16T08:07:15.763612Z TRACE drive{id=0}: quinn_proto::connection: got Data packet (33 bytes) from 127.0.0.1:60003 using id a4ec05752f480a88
2024-01-16T08:07:15.763633Z TRACE drive{id=0}:recv{space=Data pn=18801}: quinn_proto::connection: got frame StreamDataBlocked { id: StreamId(4), offset: 1250000 }
2024-01-16T08:07:15.763647Z DEBUG drive{id=0}:recv{space=Data pn=18801}:frame{ty=STREAM_DATA_BLOCKED}: quinn_proto::connection: peer claims to be blocked at stream level stream=client bidirectional stream 1 offset=1250000
2024-01-16T08:07:15.763754Z TRACE drive{id=0}:send{space=Data pn=154}: quinn_proto::connection: ACK ArrayRangeSet([91..18802]), Delay = 0us
2024-01-16T08:07:15.763778Z TRACE drive{id=0}: quinn_proto::connection: sending 36 bytes in 1 datagrams
2024-01-16T08:07:24.689968Z TRACE drive{id=0}: quinn_proto::connection: timeout timer=Idle
2024-01-16T08:07:24.690002Z TRACE drive{id=0}: quinn_proto::connection: connection closed
```

**Comment by djc** [comment]
Hmm, that does seem surprising. What do you mean by "after the tunnel is established" exactly? Do you mean the QUIC connection or some tunnel at a different level?

**Comment by paxsonsa** [comment]
Yea, QUIC connection is established here
```rust
let incoming = endpoint.accept().await.unwrap();
    let conn = incoming.await.unwrap();
    println!(
        "[server] connection accepted: addr={}",
        conn.remote_address()
    );
    conn.set_max_concurrent_bi_streams(16_u8.into());
```

**Comment by djc** [comment]
So this seems pretty surprising to me but I also have a lot of context paged out. Here's what we know from your log:

1. Got stream frame traces are from https://github.com/quinn-rs/quinn/blob/main/quinn-proto/src/connection/mod.rs#L2538
2. The same function will eventually call `Streams::received()` on https://github.com/quinn-rs/quinn/blob/main/quinn-proto/src/connection/mod.rs#L2576

What's supposed to happen soon after is that `Streams:poll()` yields `StreamEvent::Opened` (from https://github.com/quinn-rs/quinn/blob/main/quinn-proto/src/connection/streams/state.rs#L747). This would happen because `Streams::on_stream_frame()` sets `self.opened[dir]` to `true` (https://github.com/quinn-rs/quinn/blob/main/quinn-proto/src/connection/streams/state.rs#L601), but whether that happens depends on the value of `self.next_remote[dir]`.

I now wonder whether #1685 regressed this somehow?

**Comment by paxsonsa** [comment]
I'll take a peek and add some more tracing to see if I can help track that down! Thanks for the initial triage

**Comment by paxsonsa** [comment]
Just adding a simplified version of the swift client to this issue for reference. I don't expect anyone to debug this code if there is an issue but folk with macOS can reproduce (or not reproduce) the result. To work with this create a new macOS command line tool project. This should send packets every second once the main connection (stream in QUIC vernacular)
```swift
//
//  main.swift
//  QuicTool
//

import Foundation
import Network

func createMessage(version: UInt8, messageType: UInt8, message: String) -> Data {
    let messageData = message.data(using: .utf8) ?? Data()
    let length = UInt32(messageData.count)
    
    var data = Data()
    data.append(version)
    data.append(messageType)

    // Convert length to 4 bytes and append (big-endian format)
    let bigEndianLength = length.bigEndian
    data.append(contentsOf: withUnsafeBytes(of: bigEndianLength) { Array($0) })

    // Append 2 bytes of padding for 8-byte alignment
    data.append(Data(repeating: 0, count: 2))
    
    // Add Message Data.
    data.append(messageData)
    return data
}

let queue = DispatchQueue(label: "quic", qos: .userInteractive)

// Create Inital Options for the tunnel.
// This is using an insecure connection as this operation is meant to be local network.
let endpoint = NWEndpoint.hostPort(host: "127.0.0.1", port: .init(integerLiteral: 4567))
let options =  NWProtocolQUIC.Options(alpn: ["demo"])

// Insecure
sec_protocol_options_set_verify_block(options.securityProtocolOptions, { (sec_protocol_metadata, sec_trust, sec_protocol_verify_complete) in
    sec_protocol_verify_complete(true)
}, queue)
let parameters = NWParameters(quic: options)

// Create a new multiplexed connection
let descriptor = NWMultiplexGroup(to: endpoint)
let group = NWConnectionGroup(with: descriptor, using: parameters)
var mainConn: NWConnection? = nil

// Here we are establishing a state handler for when the connection to the
// the server is neogiated and "ready". Once its ready we want to open a new
// stream using the group with the options set.
//
// This is the main location of the issue we are seeing where the stream is
// established and the data is sent but never updated.
group.stateUpdateHandler = { newState in
    print("group state: \(newState)")
    switch newState {
        
    // Once the tunnel is established, create a new stream with bidirectional parameters.
    case .ready:
        print("Connected using QUIC!")
        

        // I assume I have to create a new set of options to
        // define the stream type.
        let options =  NWProtocolQUIC.Options(alpn: ["demo"])
        options.direction = .unidirectional
        mainConn = group.extract()! // force unwrap
        
        mainConn?.stateUpdateHandler = { state in
                print("Main Connection State: \(state)")
            switch state {
            case .ready:
                
                // Once the connection is ready, lets send some sweet data sauce.
                let version: UInt8 = 1
                let messageType: UInt8 = 1
                let message = "hello, I am from the multiplex group ready."
                let messageData = createMessage(version: version, messageType: messageType, message: message)
                
                mainConn?.send(content: messageData, isComplete: true, completion: .contentProcessed({ sendError in
                    if let error = sendError {
                        print("There was an error sending data: \(error)")
                    } else {
                        print("Data was sent successfully from Main Connection.")
                    }
                }))
                
            default:
                break
            }
        }
        // Don't forget to start the connection.
        mainConn?.start(queue: queue)
    default:
        break
    }
}


// Receive new incoming streams initiated by the remote endpoint
group.newConnectionHandler = { conn in
    print("New Connection: \(conn)")

  // Set state update handler on incoming stream
    conn.stateUpdateHandler = { newState in
      print("newState: \(newState) for \(conn)")
      switch newState {
      case .ready:
          let version: UInt8 = 1
          let messageType: UInt8 = 2 // Different Message Type for Remote initiated stream for debugging.
          let message = "hello, I am from the new stream from the remote"
          let messageData = createMessage(version: version, messageType: messageType, message: message)
          print("Remote stream send: \(messageData)")
          // Once the connection is ready send some delicious data sauce.
          conn.send(content: messageData, completion: .contentProcessed({ sendError in
              if let error = sendError {
                  print("There was an error sending data: \(error)")
              } else {
                  print("Data was sent successfully.")
              }
          }))

      default:
          break
      }
  }
  // Start the incoming stream
    conn.start(queue: queue)
}

// Start the group with callback queue
group.start(queue: queue)

print("running....")
while true {
    switch mainConn?.state {
    case .ready:
        // Once the connection is ready, lets send some sweet data sauce.
        let version: UInt8 = 1
        let messageType: UInt8 = 1
        let message = "hello, im from the main loop"
        let messageData = createMessage(version: version, messageType: messageType, message: message)
        print("Local Stream Send: \(messageData)")

        mainConn?.send(content: messageData, isComplete: true, completion: .contentProcessed({ sendError in
            if let error = sendError {
                print("There was an error sending data: \(error)")
            }
        }))
        sleep(1)
    default:
        continue
    }
}
```

**Comment by djc** [comment]
Is there a command-line invocation for the Swift toolchain that makes it easy to run this? I have a Mac but I've never really used Xcode.

**Comment by paxsonsa** [comment]
Let me whip something together.

**Comment by paxsonsa** [comment]
Try this:
https://github.com/paxsonsa/quic-swift-demo

```bash
swift run -v
```

**Comment by Ralith** [comment]
Thanks for the detailed report! I agree that this looks like a Quinn bug. We're clearly receiving stream data but failing to unblock your application code. From your original post, I understand that `accept_bi` succeeds, which means the `StreamEvent::Opened` event is being successfully propagated. For some reason, [the `ReadExact` future](https://github.com/quinn-rs/quinn/blob/20dff915e1feaf293a739e68dc2c6ea2c6bbca09/quinn/src/recv_stream.rs#L536-L553) isn't becoming ready, even though you've received far more than the required amount of data. That suggests a bug in its `poll` implementation, or `quinn::RecvStream::poll_read`, or `quinn::RecvStream::poll_read_generic`, or in the `Chunks` helper from `quinn_proto::RecvStream::read`.

Could you attach a debugger to your server with a breakpoint on `ReadExact::poll`, then step through it to see what path it returns `Poll::Pending` through?

**Comment by paxsonsa** [comment]
@Ralith The other way, `open_bi` works (eg initiating from the server) `accept_bi` does not. Sorry for the confusion

BUT it's the same thing and I have enough info to go looking deeper, let me whip that up and get back to you

**Comment by Ralith** [comment]
> @Ralith The other way, open_bi works (eg initiating from the server) accept_bi does not. Sorry for the confusion

I meant that, according to your report, the `accept_bi()` operation completes successfully, and that the program then later times out during `read_exact`. Hence, we know the problem isn't in accepting the stream. The fact that using a client-initiated stream instead succeeds is interesting, and suggestive of a weird state machine bug in the proto layer, but maybe it's just timing sensitive.

**Comment by paxsonsa** [comment]
This really interesting....

TLDR; After going done the path @Ralith mentioned I found that the stream was blocked and never being awoken by the connection. I add some instrumentation and tracing to the connection events and behold! **2 Streams were made**. 

I am still not clear about what is happening, how can two stream be open, is the Network.framework not doing the right thing and actually waiting for the stream to be accepted before giving me a valid `.ready` state or is quinn proactively responding to Open events? Both?

Here is a log with some of that instrumetation
```
Event: Connected
[server] connection accepted: addr=127.0.0.1:63255
waiting for bi stream
2024-01-18T10:22:52.410294Z TRACE drive{id=0}: quinn_proto::connection: got Data packet (31 bytes) from 127.0.0.1:63255 using id b21fb38c4754fb33
2024-01-18T10:22:52.410357Z TRACE drive{id=0}:recv{space=Data pn=0}: quinn_proto::connection: got frame Ack(Ack { largest: 2, delay: 0, ecn: None, ranges: "[0..=2]" })
2024-01-18T10:22:52.410393Z TRACE drive{id=0}:recv{space=Data pn=0}:frame{ty=ACK}: quinn_proto::connection::mtud: new MTU detected current_mtu=1326
2024-01-18T10:22:52.410440Z TRACE drive{id=0}:send{space=Data pn=3}: quinn_proto::connection::packet_builder: PADDING * 1370
2024-01-18T10:22:52.410459Z TRACE drive{id=0}: quinn_proto::connection: writing MTUD probe probe_size=1389
2024-01-18T10:22:52.410467Z TRACE drive{id=0}: quinn_proto::connection: sending 1389 bytes in 1 datagrams
2024-01-18T10:22:52.411138Z TRACE drive{id=0}: quinn_proto::connection: got Data packet (80 bytes) from 127.0.0.1:63255 using id b21fb38c4754fb33
2024-01-18T10:22:52.411171Z TRACE drive{id=0}:recv{space=Data pn=1}: quinn_proto::connection: got stream frame id=client bidirectional stream 1 offset=0 len=51 fin=false
chunk assmeble insert()
insert() post-dedupe data.len() = 0
insert() data.len() = 1
2024-01-18T10:22:52.411205Z TRACE drive{id=0}: quinn_proto::connection: got Data packet (66 bytes) from 127.0.0.1:63255 using id b21fb38c4754fb33
2024-01-18T10:22:52.411228Z TRACE drive{id=0}:recv{space=Data pn=2}: quinn_proto::connection: got stream frame id=client bidirectional stream 1 offset=51 len=36 fin=false
chunk assmeble insert()
insert() post-dedupe data.len() = 1
insert() data.len() = 2
2024-01-18T10:22:52.411264Z TRACE drive{id=0}:send{space=Data pn=4}: quinn_proto::connection: ACK ArrayRangeSet([0..3]), Delay = 0us
2024-01-18T10:22:52.411284Z TRACE drive{id=0}: quinn_proto::connection: sending 26 bytes in 1 datagrams
Event: Stream(Opened { dir: Bi })
Event: Stream(Readable { id: StreamId(4) })
Readable: client bidirectional stream 1
reading from recv
ReadExact::poll()
ReadExact::poll iter -- remaining 8
poll_read -- remaining 8
poll_read_generic -- got conn lock
poll_read_generic -- got chunks
poll_read_generic :: read_fn -- remaining 8
chunk.next() state=Readable(Recv { state: Recv { size: None }, assembler: Assembler { state: Ordered, data: [], buffered: 0, allocated: 0, bytes_read: 0, end: 0 }, sent_max_stream_data: 1250000, end: 0, stopped: false })
chunk assmeble read()
chunk.next() assembler.read() was None: Recv { state: Recv { size: None }, assembler: Assembler { state: Ordered, data: [], buffered: 0, allocated: 0, bytes_read: 0, end: 0 }, sent_max_stream_data: 1250000, end: 0, stopped: false }
chunk.next() RecvState::Recv: size=None
poll_read_generic -- read_fn done
poll_read_generic -- read_fn status: Failed(Blocked)
poll_read_generic -- blocked on stream: StreamId(0) -- pending.
poll_read -- returning Pending
ReadExact::poll stream poll_read returning with pending
```
Where you see `Event` that is an event being emitted by the connection `State::forward_app_events` and you'll see at some point a there is a `StreamId(0)` and a `StreamId(4)`. Not really sure how and then, in the tracing logs you also see a `stream 1` ????

🤯 

Now the fix here is easy to get the example working: call `accept_bi()` twice. I did that and it works as expected.


**Root Issue**
Without accepting the stream, why are we getting data and getting a ready state on the client side for that second stream? Not sure what the spec says about that but I assume that the stream state should not be valid until we accept_bi?

Where can I go next to help you out with this? I might need to dive into the RFC to be helpful.

**Comment by Ralith** [comment]
Thanks for the details! Two streams being open is the key. This is most likely a bug in your application logic after all, but maybe a surprising interaction we (and Quinn implementers in general) could document in more detail.

The key thing you noticed here is that your read operation is blocked on `StreamId(0)`, aka "client bidirectional stream 0", whereas you're receiving data for `StreamId(4)`, aka "client bidirectional stream 1". From your logs, it appears that your server has only received frames regarding "client bidirectional stream 1". However, per RFC9000:

> [A stream ID that is used out of order results in all streams of that type with lower-numbered stream IDs also being opened.](https://www.rfc-editor.org/rfc/rfc9000.html#name-stream-types-and-identifier)

Receiving data for that stream implicitly opened the lower-numbered stream. `accept_bi` yields streams in the order that they are opened, which is by definition ascending numerical order. Because you haven't received any data on "client bidirectional stream 0", reading from it does not complete. Either your client should either write to the first stream it opens rather than the second, or your server should spawn off tasks to read from multiple streams concurrently.

If you don't intend for your client to open multiple streams, there might be something funny going on with Apple's QUIC implementation, but I would be very certain of your client's behavior first. I don't know Swift or Apple's QUIC API well enough to easily audit it myself.

**Comment by Ralith** [comment]
> Without accepting the stream, why are we getting data and getting a ready state on the client side for that second stream? Not sure what the spec says about that but I assume that the stream state should not be valid until we accept_bi?

For a little more general background here: QUIC expects you to be using multiple streams concurrently, so by default peers are given permission to open a decent number of streams unilaterally. Even if you only intend to use streams one at a time, this ensures that the sender doesn't have to sit idle for a full round trip each time it wants to make a new one.

You could adjust Quinn's settings to permit only one stream at a time, but that's probably not what you really want.

**Comment by paxsonsa** [comment]
Perfect! Thank you for the help and detailed explanation here!

You all rock! Keep up the awesome! 

Closing!

**Comment by objectivecosta** [comment]
@paxsonsa I encountered [a similar issue recently](https://forums.developer.apple.com/forums/thread/750808), and it stems from the fact that's it's easy to open client streams on `NWConnectionGroup` but there doesn't seem to be a way to send/receive on stream ID `0`. I was utilizing quiche instead of quinn, but same result. Safe to say that this is not backend-implementation-dependent.
